### DISCUSSION TITLE
Add pooling, semaphores, latches, and mutexes

### DISCUSSION THREAD

**Discussion by aleclarson** [radashi-org/radashi#222] [Ideas]
**Pooling:** Manages reusable objects to reduce creation/destruction overhead. Useful for expensive resources like database connections or complex data structures.

**Semaphores:** Limit concurrent asynchronous operations. Control simultaneous API calls or database queries to prevent resource exhaustion or rate limit breaches.

**Latch:** Coordinates multiple asynchronous operations, allowing you to wait until all are complete. Initializes with a count, decrements as operations finish, and resolves a promise when the count reaches zero.

**Mutex:** Ensures non-overlapping access to shared resources in asynchronous code. Implements critical sections for async operations, preventing interference when multiple async functions run concurrently.

These tools optimize resource management and control flow in JavaScript's asynchronous, event-driven environment.

We are currently taking suggestions on which popular NPM libraries to model our simple reimplementation on. Our goal is to mimic the API of a popular NPM library for each of these, so you can always upgrade to their more feature-rich implementations when/if you need it.

**Comment by aleclarson** [comment]
When creating this post, I had the wrong idea of what a semaphore is (I was really thinking of a latch, apparently). I believe #145 covers the use cases of a semaphore, though I could be wrong. If anyone has more familiarity, feel free to confirm or deny.

I suppose a semaphore API could look more like a class, rather than the currying style of `bottleneck` in #145. A good example of that is Shopify's semaphore package: [`@shopify/semaphore`](https://www.npmjs.com/package/@shopify/semaphore). They claim in their readme that a semaphore can be used for mutex behavior as well, so we may not need a separate mutex implementation.

**Comment by adamhamlin** [comment]
Re: latch -- What kind of functionality are you thinking about beyond what `Promise.all` or [Promise.allSettled](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled) already provides, a la `await Promise.allSettled(myPromises)`?

**Reply by aleclarson** [reply]
Maybe a “dynamic” latch (where the count is not known beforehand) could be useful? For example, when scraping a web page, you may recursively scan linked pages, but you can't know if you'll find any until you do. To be fair, that's not the best example, since you'd probably want a purpose-built library for web scraping.

I agree that `Promise.all` and `allSettled` are good enough for known counts. I don't know of a use case where being able to pass around the latch would be necessary or substantially better, but that's the advantage I see over using `Promise.all` etc.

**Comment by hugo082** [comment]
Here is some related functions proposal. I would love to get your feedbacks in order to send PRs.

Proposal inspired from [@shopify/semaphore](https://www.npmjs.com/package/@shopify/semaphore), [async-mutex](https://github.com/DirtyHairy/async-mutex) & rust [async_weighted_semaphore](https://docs.rs/async-weighted-semaphore/latest/async_weighted_semaphore/index.html).

Let me know that are the next steps.

#### `createSemaphore` proposal
```ts
import * as _ from 'radashi'

const semaphore = _.createSemaphore(/* size: */ 2)

// available properties / methods
const release = await semaphore.acquire({ weight: 1, priority: 1 })
semaphore.isLocked()
release()

// ran one by one
await _.all([
	semaphore.exclusive(async () => /* ... */, { weight: 1, priority: 1 }),
	semaphore.exclusive(async () => /* ... */),
])
```

#### `createMutex` proposal
```ts
import * as _ from 'radashi'

// same as createSemaphore(1)
const mutex = _.createMutex()
```

#### `exclusive` proposal

```ts
import * as _ from 'radashi'

// create a mutex (semaphore = 1) around the function
const exclusiveFn = _.exclusive(async () => /* ... */)

// ran one by one
await _.all([
	exclusiveFn(),
	exclusiveFn()
])
```

**Reply by aleclarson** [reply]
Thank you for your proposal.

I'll need some additional context in order to evaluate your proposal:
1. Are these functions as simple as possible while still fulfilling their primary use case?
2. Can we align the proposed APIs with an existing library, to ensure that users can easily migrate to a more advanced implementation if necessary?
3. Have you tried implementing these functions yet?
4. Which use cases have you considered before landing on this design?

Feel free to reply with any questions or clarifications. 👍 

Also, welcome to the Radashi community! It's great to have you. 😄

**Reply by hugo082** [reply]
#### 1. Are these functions as simple as possible while still fulfilling their primary use case?

I feel like they are. They takes only the `size | limit` as option and `priority` & `weight` (optionally) to request a lock.

#### 2. Can we align the proposed APIs with an existing library, to ensure that users can easily migrate to a more advanced implementation if necessary?

The main difference with both libraries I mentioned ([@shopify/semaphore](https://www.npmjs.com/package/@shopify/semaphore), [async-mutex](https://github.com/DirtyHairy/async-mutex)) is that we hide the `new Semaphore` behind the `createMutex` & `createSemaphore`.
This is to follow the radashi API which expose functions only, but we can change to expose the Semaphore class instead.

#### 3. Have you tried implementing these functions yet?

Not yet, but the libraries mentioned are pretty clear on the implementation. I prefer to wait for the validation of this proposal 🙃

#### 4. Which use cases have you considered before landing on this design?

My personal use-case is related to [Temporal Workflow updates handling](https://docs.temporal.io/develop/typescript/message-passing#control-handler-concurrency) but Semaphore is a well known design pattern.
> https://github.com/DirtyHairy/async-mutex
> While Javascript is strictly single-threaded, the asynchronous nature of its execution model allows for race conditions that require similar synchronization primitives. Consider for example a library communicating with a web worker that needs to exchange several subsequent messages with the worker in order to achieve a task. As these messages are exchanged in an asynchronous manner, it is perfectly possible that the library is called again during this process. Depending on the way state is handled during the async process, this will lead to race conditions that are hard to fix and even harder to track down.

**Reply by aleclarson** [reply]
Hey @hugo082, sorry the delayed reply!

Would the `bottleneck` function implemented in #145 work for your use case?

**`createSemaphore` alternative**
```ts
import * as _ from 'radashi'

const exclusive = _.bottleneck({ concurrency: 2 }, fn => fn())

// runs two at a time
await _.all([
	exclusive(async () => /* ... */),
	exclusive(async () => /* ... */),
	exclusive(async () => /* ... */),
	exclusive(async () => /* ... */),
])
```

**`exclusive` alternative**
```ts
import * as _ from 'radashi'

// create a mutex (semaphore = 1) around the function
const exclusiveFn = _.bottleneck({ concurrency: 1 }, async () => /* ... */)

// ran one by one
await _.all([
	exclusiveFn(),
	exclusiveFn()
])
```

The only missing features are the `weight` and `priority`. So maybe we could use a function that implements those on top of bottleneck?

Let me know your thoughts 😄

**Reply by aleclarson** [reply]
Thought about `bottleneck` some more. Maybe it's best that we have a dedicated function for semaphore purposes, so it can enjoy a lighter implementation. Then again, I don't want too many functions with a similar use case. 🤔 

Here's the API I'm thinking we should go with (feedback welcome). I'm not sold on the naming (`withSemaphore`), but since it's returning a function, I don't think `createSemaphore` is as good.

```ts
import { sleep } from 'radashi'

const start = Date.now()
const elapsed = () => Date.now() - start

// Signature #1
// When just a number is passed, it creates a queue function.
const queue = withSemaphore(1)

queue(async () => {
  console.log(elapsed()) // Logs 0
  await sleep(100)
})

queue(async () => {
  console.log(elapsed()) // Logs 100
  await sleep(100)
})

// Signature #2
// When a callback is also passed, it's bound to the queue function.
const logTimeSoon = withSemaphore(1, async () => {
  console.log(elapsed())
  await sleep(100)
})

logTimeSoon() // => Logs 0
logTimeSoon() // => Logs 100
```

As far as weight/priority logic, I think we should save that for a higher level function that wraps the semaphore function. I've implemented a proof of concept here:

```ts

function withWeightedSemaphore({
  maxWeight,
  concurrency = Number.POSITIVE_INFINITY,
  strict,
}: {
  /**
   * The maximum weight that can be run concurrently.
   */
  maxWeight: number
  /**
   * The maximum number of tasks that can be run concurrently.
   *
   * @default Infinity
   */
  concurrency?: number
  /**
   * Instead of jumping over tasks due to weight concerns, wait for
   * enough vacancy to become available.
   *
   * @default false
   */
  strict?: boolean
}) {
  // Use a semaphore if concurrency is limited.
  const apply = Number.isFinite(concurrency)
    ? withSemaphore(concurrency)
    : fn => fn()

  const queue: {
    task: () => Promise<any>
    weight: number
    priority: number
    resolve: (value: Promise<any>) => void
  }[] = []

  let currentWeight = 0

  // Call the next task with a small enough weight.
  const shift = () => {
    for (let i = 0; i < queue.length; i++) {
      const { task, weight, resolve } = queue[i]

      if (currentWeight + weight <= maxWeight) {
        currentWeight += weight
        queue.splice(i, 1)

        return resolve(
          task().finally(() => {
            currentWeight -= weight

            // Since shift may be a no-op due to weight concerns, call
            // apply to ensure the next task is called.
            if (queue.length) {
              apply(shift)
            }
          }),
        )
      }
      if (strict) {
        return
      }
    }
  }

  return <T>(
    {
      weight,
      priority = 0,
    }: {
      weight: number
      priority?: number
    },
    task: () => Promise<T>,
  ) =>
    new Promise<T>(resolve => {
      if (weight > maxWeight) {
        throw new Error(`Weight ${weight} exceeds maxWeight ${maxWeight}`)
      }
      let index = queue.findIndex(item => item.priority > priority)
      if (index === -1) {
        index = queue.length
      }
      queue.splice(index, 0, {
        task,
        weight,
        priority,
        resolve,
      })
      apply(shift)
    })
}

const weightedLog = withWeightedSemaphore({
  concurrency: 2,
  maxWeight: 2,
  strict: true, // Don't prioritize small tasks.
})

weightedLog({ weight: 1 }, async () => {
  console.log('A:', elapsed())
  await sleep(100)
})

// Even though concurrency is 2, this won't run until the first task is complete,
// because its weight equals the maxWeight.
weightedLog({ weight: 2 }, async () => {
  console.log('B:', elapsed())
  await sleep(100)
})

// Plot twist. We're queueing a task with higher priority than the
// previous task that was queued. Its weight is only 1, so it will run
// in parallel with the first task.
weightedLog({ weight: 1, priority: 1 }, async () => {
  console.log('C:', elapsed())
  await sleep(100)
})

// Logs:
// A: 0
// C: 0
// B: 100
```

Let me know if you'd like to spearhead the PR. 👍

**Reply by hugo082** [reply]
IMO it's useful to keep the `createSemaphore` & `createMutex` APIs to provide access to custom management (`acquire` / `isLocked` / `release` methods) and making it behave similarly to standard semaphore implementations.

But I like your `withSemaphore` double signature proposal, maybe we can implement it in place of the `exclusive` function & in addition of `create*` APIs?

Regarding the `withWeightedSemaphore`, IMO it's useless to expose `maxWeight` (which doesn't make sense to me - we should use the `concurrency`/`capacity` option) & `strict` (let's keep the behavior simple, FIFO, and maybe implement priority later).

```ts
const limited = withSemaphore(1);
limited(() => { /* ... */ })

const limited = withSemaphore(1, () => { /* ... */ });
limited();

const limited = withSemaphore({ capacity: 2 })
limited({ weight: 2 });

const limited = withSemaphore({ capacity: 2 }, () => { /* ... */ })
limited({ weight: 2 });

function withSemaphore(capacity: number): Function;
function withSemaphore(options: Options): Function;
function withSemaphore(capacity: number, fn: Function): Function;
function withSemaphore(options: Options, fn: Function): Function;
```

Let me know your thoughts :)

**Reply by aleclarson** [reply]
> IMO it's useless to expose `maxWeight` (which doesn't make sense to me - we should use the `concurrency`/`capacity` option) & `strict` (let's keep the behavior simple, FIFO, and maybe implement priority later).

I will trust your opinion on this matter 👍 

> it's useful to keep the `createSemaphore` & `createMutex` APIs to provide access to custom management

I'd like to avoid `createSemaphore` and `createMutex` if we can help it. For example, you can wrap `withSemaphore` to get similar behavior, and call it `withInvertedSemaphore` or something.

Here I'm wrapping `withSemaphore` to create a `withMutex` function that doesn't take a job function, allowing the caller to essentially *be* the job. Unlocking is then delegated to the caller. It's questionable whether this is a pattern worth including in Radashi, as it's more error-prone than `withSemaphore`'s automatic unlocking.

```ts
/// <reference lib="esnext" />

declare function withSemaphore(capacity: number): {
  <T>(fn: () => Promise<T>): Promise<T>
  hasCapacity(weight?: number): boolean
}

type NotUndefined = {} | null

export type MutexResult<T> = T extends never ? () => void : [T, () => void]

export type Mutex<T> = {
  (): Promise<MutexResult<T>>
  isAvailable(): boolean
}

/**
 * Create a function that requests exclusive access to a resource.
 */
export function withMutex<T extends NotUndefined = never>(
  resource?: T,
): Mutex<T> {
  const queue = withSemaphore(1)

  async function mutex(): Promise<MutexResult<T>> {
    return new Promise(next => {
      queue(() => {
        const lock = Promise.withResolvers<void>()
        next(
          (resource !== undefined
            ? ([resource, lock.resolve] as const)
            : lock.resolve) as MutexResult<T>,
        )
        return lock.promise
      })
    })
  }

  mutex.isAvailable = queue.hasCapacity
  return mutex
}

const getThing = withMutex(new Thing())

const [thing, releaseThing] = await getThing()
getThing.isAvailable() // => false
await thing.doThing()
releaseThing()
```

I also wonder if there's a common use case for semaphores with “manual lock management” methods, or if it's usually a mutex that requires such capability?

You *could* adapt `withMutex` above to accept a capacity argument. If we can avoid bloating `withSemaphore` with this responsibility, I would like to do so.

**Reply by hugo082** [reply]
Ok I will take some time during the week to send a PR that implement withSemaphore, let's iterate on it 👍

**Reply by aleclarson** [reply]
Hey @hugo082! It's been about 12 days, so I'm checking in. Let me know your status so I can plan accordingly. Thanks! :)

**Reply by hugo082** [reply]
Hi @aleclarson! Sorry for the late update but here is the PR: https://github.com/radashi-org/radashi/pull/331 😄

**Reply by aleclarson** [reply]
Hey @hugo082, thanks for opening the PR! It's great to have your help. I wrote my own implementation for my own use, before you had opened #331, so I opened #335 for us to compare our approaches and come to consensus on what should be merged. Let me know what you think 👍

**Reply by hugo082** [reply]
Hey @aleclarson, I just checked your PR.
IMO it's a good thing to match the standard so users can quickly find the function & know that the function is doing.
Adding a manual management on `acquire/release` allows to deal with async locking mechanism - .e.g. the release comes from an event. I agree, it's more error prone but I don't think this is the responsibility of the lib to ensure the code quality.
About the `getRunning` it is always useful to be able to get this info for debugging, logging or whatever. I'm ok to add the `hasCapacity`, could be useful too.
Note that the `withMutex` is equivalent to `withSemaphore({ capacity: 1 }, ...)` but expose different usage interface `isLocked/acquire/release`
